include/exadg/operators/interior_penalty_parameter.h

```diff
@@ -25,67 +25,118 @@
+  };
 
-  QGauss<dim - 1>   face_quadrature(degree + 1);
-  FEFaceValues<dim> fe_face_values(mapping, fe, face_quadrature, update_JxW_values);
+  FaceIntegrator<dim, 1, Number> face_integrator(matrix_free, true, dof_index);
 
-  for(unsigned int i = 0; i < n_cells; ++i)
+  for(unsigned int face = 0;
+      face < matrix_free.n_inner_face_batches() + matrix_free.n_boundary_face_batches();
+      ++face)
   {
-    for(unsigned int v = 0; v < matrix_free.n_active_entries_per_cell_batch(i); ++v)
+    face_integrator.reinit(face);
+    VectorizedArray<Number> face_area = 0.0;
+    for(unsigned int q = 0; q < face_integrator.n_q_points; ++q)
+      face_area += face_integrator.JxW(q);
+
+    bool const is_interior_face = face < matrix_free.n_inner_face_batches();
+
+    // For interior faces, the face area is weighted by a factor of 0.5
+    // according to the formula by "K. Hillewaert, Development of the
+    // Discontinuous Galerkin Method for High-Resolution, Large Scale CFD
+    // and Acoustics in Industrial Geometries, Ph.D. thesis, Univ. de
+    // Louvain, 2013."
+    if(is_interior_face)
+      face_area = face_area * 0.5;
+
+    // Write the result to the vector entry corresponding to the respective
+    // cell slot. For boundary faces, we write it to only one cell,
+    // otherwise to both adjacent cells of the face.
+    for(unsigned int v = 0; v < matrix_free.n_active_entries_per_face_batch(face); ++v)
     {
-      typename DoFHandler<dim>::cell_iterator cell = matrix_free.get_cell_iterator(i, v, dof_index);
-      fe_values.reinit(cell);
-
-      // calculate cell volume
-      Number volume = 0;
-      for(unsigned int q = 0; q < quadrature.size(); ++q)
-      {
-        volume += fe_values.JxW(q);
-      }
-
-      // calculate surface area
-      Number surface_area = 0;
-      for(const unsigned int f : cell->face_indices())
-      {
-        fe_face_values.reinit(cell, f);
-        Number const factor = (cell->at_boundary(f) && !cell->has_periodic_neighbor(f)) ? 1. : 0.5;
-        for(unsigned int q = 0; q < face_quadrature.size(); ++q)
-        {
-          surface_area += fe_face_values.JxW(q) * factor;
-        }
-      }
-
-      array_penalty_parameter[i][v] = surface_area / volume;
+      surface_areas(get_cell_index(matrix_free.get_face_iterator(face, v, true).first)) +=
+        face_area[v];
+
+      if(is_interior_face)
+        surface_areas(get_cell_index(matrix_free.get_face_iterator(face, v, false).first)) +=
+          face_area[v];
     }
   }
+
+  surface_areas.compress(VectorOperation::add);
+  surface_areas.update_ghost_values();
+
+  // As a second step, we use a CellInterator to compute the volume and then
+  // combine it with the surface areas computed before
+  unsigned int const n_cells = matrix_free.n_cell_batches() + matrix_free.n_ghost_cell_batches();
+  array_penalty_parameter.resize(n_cells);
+  CellIntegrator<dim, 1, Number> integrator(matrix_free, dof_index);
```

Comment:
quad_index does not matter?

Reply:
I also had this question. I think we should extend the interface of this function (and remove the default value for dof_index). In the old code version, the quadrature rule was constructed inside the function.

Reply:
Let's keep this to a follow-up PR!

Reply:
Why? This PR might introduce a bug!?

Reply:
Apart from matrix-based vs. matrix-free evaluation, the code might do something different now

Reply:
> This PR might introduce a bug!?

You are right. But you might need to help Martin. I quickly take a look where the function is called and at some of those places no quad_index is available (e.g., https://github.com/exadg/exadg/blob/7af16c94168fc0126315d55c038ea55ad9e0fe47/include/exadg/incompressible_navier_stokes/spatial_discretization/operators/viscous_operator.h#L64-L67). Also one needs to decide which quad index to use which might be not unique.

Reply:
Yes, we can introduce that parameter, but the follow-up question is what quad index we should be choosing. Before, we used `degree + 1`, while the natural choice now is the one in the evaluator context where we're using it (which is often `degree + 1`). We might have minor deviations in behavior between the previous version and current one in those applications that do not use `degree + 1` quadrature points (maybe we do everywhere now, but in the future we might decide differently). Anyway, it shouldn't matter much given that nature of the information we're computing: an approximation of the area/volume. The underlying formula is by no means exact, so a minor quadrature error up or down is irrelevant, and often we either get the exact result or a reasonable estimate already with a single quadrature point. In all our use cases, the number of quadrature points scales with the polynomial degree (i.e., also the mapping degree in the deformed case), so we can expect very small quadrature errors.

Reply:
> Apart from matrix-based vs. matrix-free evaluation, the code might do something different now

I am confused by the terms *matrix-based* vs *matrix-free*, but I assume you mean `FEValues` vs `FEEvaluation`? In principle, they should be equivalent apart from round-off errors. But the main issue I see is that I could have introduced a bug in what I execute here, as the parameter is hard to verify. I tested several Poisson problems, but it might be good to keep an eye on it.

Reply:
I now added this with the separate commit ab0ff36. It is not particularly pretty to pull it through all interfaces, but this might be needed at some point anyway (see e.g. #128).

Reply:
> I am confused by the terms matrix-based vs matrix-free, but I assume you mean FEValues vs FEEvaluation

Yes. I have written this very quickly yesterday.

Reply:
> It is not particularly pretty to pull it through all interfaces, but this might be needed at some point anyway (see e.g. #128).

We might want to refactor later, but the present PR should make efforts to not change the numerics.

Reply:
> You are right. But you might need to help Martin. I quickly take a look where the function is called and at some of those places no quad_index is available

I have carefully gone through all changes and added comments where adjustments are necessary.